### ISSUE TITLE
[Bug] MariaDB custom config disregarded when upgrading from 0.27.0 to 0.28.0

### ISSUE THREAD

**Issue by markus-brln** [mariadb-operator/mariadb-operator#827]
**Documentation**
- [x] I acknowledge that I have read the relevant [documentation](https://github.com/mariadb-operator/mariadb-operator/tree/main/docs).

**Describe the bug**

Upgrading the operator and the MariaDB CR object from 0.27.0 to 0.28.0 caused the key of the config map mentioned in   `myCnfConfigMapKeyRef` to be mounted to a different file. Originally, it was mounted to `/etc/mysql/conf.d/my.cnf`. After the upgrade it becomes `/etc/mysql/conf.d/<name of configmap key>`, which un-applies the config if the configmap's key is not postfixed with the `.cnf` file extension.

We fixed it by appending the `.cnf` extension in the configmap and the `myCnfConfigMapKeyRef`. Luckily, we're on 0.29.0 already so that was possible, on 0.28.0 the field is immutable, according to my tests.

**Expected behaviour**

- Append the `.cnf` extension by itself or give an error that a certain value is expected
- Another way to mitigate these problems would be note about this required step in the  [0.28.0 upgrade guide](https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/UPGRADE_v0.0.28.md) and make it more clear what key names are allowed in the [CONFIGURATION.md](https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/CONFIGURATION.md). The `myCnf` key [mentioned](https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/CONFIGURATION.md#mycnf) there has the same problem as my `config` key name.

**Steps to reproduce the bug**

Replicate this bug by following the [0.28.0 upgrade guide](https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/UPGRADE_v0.0.28.md):
- Install mariadb operator 0.27.0
- Install `MariaDB` CR object, containing:

```yaml
apiVersion: k8s.mariadb.com/v1alpha1
kind: MariaDB
metadata:
  name: mariadb
spec:
  myCnfConfigMapKeyRef:
    name: mariadb-config
    key: config
```
- configmap with example non-default value:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: mariadb-config
data:
  config: |-
    [mysqld]
    # default 151 connections
    max_connections=300
```
- Execute `SHOW VARIABLES;` to check that `max_connections` == 300
- The custom config file is mounted as `my.cnf`, while the key of the configmap is `config` (`root@mariadb-0:/etc/mysql/conf.d# ls` -> `my.cnf`)
- Uninstall the operator version 0.27.0
- Upgrade CR object to 0.28.0 according to the upgrade guide
- Install the operator version 0.28.0
  - Triggers restart of mariadb pods
  - Execute `SHOW VARIABLES;`,  `max_connections` will have defaulted back to 151
  - If you go to the config dir, you will find a default config file and a file that is called the same as the key in the config map (`root@mariadb-0:/etc/mysql/conf.d# ls` -> `0-default.cnf config`)

**Debug information**

- Nothing unusual from the operator
- One pod shows this
```json
{"level":"debug","ts":1725444139.6744573,"logger":"galera.health","msg":"Checking Galera cluster health","controller":"statefulset","controllerGroup":"apps","controllerKind":"StatefulSet","StatefulSet":{"name":"mariadb","namespace":"gomibo-saas"},"namespace":"gomibo-saas","name":"mariadb","reconcileID":"562f82eb-b7e3-4785-831a-214871c6855e"}
```
- The others only have the init logs:
```json
{"level":"info","ts":1725437019.8687599,"logger":"setup","msg":"Watching all namespaces"}
{"level":"info","ts":1725437019.8739328,"logger":"setup","msg":"Starting manager"}
{"level":"info","ts":1725437019.8739867,"logger":"controller-runtime.metrics","msg":"Starting metrics server"}
{"level":"info","ts":1725437019.8740528,"logger":"controller-runtime.metrics","msg":"Serving metrics server","bindAddress":":8080","secure":false}
{"level":"info","ts":1725437019.874112,"msg":"starting server","kind":"health probe","addr":"[::]:8081"}
I0904 08:03:39.974911       1 leaderelection.go:250] attempting to acquire leader lease mariadb-operator/mariadb-operator.mariadb.com...
```
- In other words, the operator or the MariaDB CR object don't complain. The problems only start when you expect the mariadb custom config to be applied

**Environment details**:
- Kubernetes version: 1.28.4
- Kubernetes distribution: RKE2, microk8s on local
- mariadb-operator version: 0.28.0
- Galera agent (sidecar container): 0.0.28
- Install method: any
- Install flavor: any

**Additional context**

I feel like this PR may have something to do with the changed behavior: https://github.com/mariadb-operator/mariadb-operator/pull/487 ([release docs reference](https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/releases/RELEASE_v0.0.28_HEADER.md.gotmpl#L19))

**Comment by mmontes11** [comment]
Hey there @markus-brln ! Thanks a lot for the detailed report and the investigation.

I've updated the docs to point this out explicitly: https://github.com/mariadb-operator/mariadb-operator/commit/b0a80c98a0210cad3e31af9b44e1ba65572bf35b

I would add a validation `myCnfConfigMapKeyRef.key` in the `MariaDB` validation webhook to ensure that it ends in `.cnf` and otherwise return an error to the user:
https://github.com/mariadb-operator/mariadb-operator/blob/36afbefcefb66e0a3f10877d9e1fe0150fa1ad4b/api/v1alpha1/mariadb_webhook.go#L43

It should be pretty straightforward. Contributions are welcome!

**Comment by am6737** [comment]
Hi @markus-brln

Do you have time to fix this issue? If not, would you mind if I take a look?
